### ISSUE TITLE
15.11: ejabberdctl dump yields timeout

### ISSUE THREAD

**Issue by mathiasertl** [processone/ejabberd#866]
Since the upgrade to 15.11 I get this error:

```
root@host: ~ $ ejabberdctl dump ejabberd.dump
Failed RPC connection to the node ejabberd@titan: timeout

Commands to start an ejabberd node:
  start    Start an ejabberd node in server mode
  debug    Attach an interactive Erlang shell to a running ejabberd node
  iexdebug Attach an interactive Elixir shell to a running ejabberd node
  live     Start an ejabberd node in live (interactive) mode
  iexlive  Start an ejabberd node in live (interactive) mode, within an Elixir shell

Optional parameters when starting an ejabberd node:
  --config-dir dir   Config ejabberd:    /etc/ejabberd
  --config file      Config ejabberd:    /etc/ejabberd/ejabberd.yml
  --default file     Config ejabberdctl: /etc/default/ejabberd
  --default-dir      Config ejabberdctl: /etc/default
  --logs dir         Directory for logs: /var/log/ejabberd
  --spool dir        Database spool dir: /var/lib/ejabberd
  --node nodename    ejabberd node name: ejabberd@titan
```

starting the command with `time` yields:

```
root@host: ~ $ time ejabberdctl dump ejabberd.dump
...
real    1m0.624s
user    0m0.252s
sys     0m0.016s
```

Is the timeout configurable? Was it changed with ejabberd 15.11? Or is there a different issue?

**Comment by weiss** [comment]
> Is the timeout configurable?

No, currently it's [hard-coded][1].

> Or is there a different issue?

We're running into the same timeout since dumping our MAM table takes too long.  (The dump succeeds despite the timeout message, but this should of course be fixed.)

[1]: https://github.com/processone/ejabberd/blob/0a846d03bf4c1a07a1b2bfcff9e81c0785533598/src/ejabberd_ctl.erl#L83

**Comment by mathiasertl** [comment]
The usefulness of the `dump` command is of course quite limited, especially since you'd typically want to do something (compress, etc.) with the dump after it's completed. But thanks for the pointer, that way I can at least increase the dump timeout in my own patch.

**Comment by mathiasertl** [comment]
... but to solve this problem: Maybe an environment variable in ejabberdctl.cfg?

**Comment by mathiasertl** [comment]
I'm afraid there might be a different issue. I just deployed a fix that increases the timeout to ten minutes, and I still get the timeout. I didn't get this timeout at all with 15.09, it started the moment I deployed ejabberd 15.11.

**Comment by mathiasertl** [comment]
Maybe this is somehow related:

```
root@host: /tmp $ ls -sh ejabberd.dump; sleep 10s; ls -sh ejabberd.dump
74M ejabberd.dump
96M ejabberd.dump
root@host: /tmp $ ls -sh ejabberd.dump; sleep 10s; ls -sh ejabberd.dump
309M ejabberd.dump
326M ejabberd.dump
```
This means a staggering 2.2 MB/s and 1.7 MB/s respectively. I'm not entirely sure, but I think this was a lot higher in the past.

**Comment by prefiks** [comment]
Before 15.9 there was no timeout, i believe there was no recent changes in dump code that should slow it down. You can disable this timeout and go back to what was in 15.09 by removing last argument in rpc:call.

**Comment by prefiks** [comment]
In commit 7a69dab5963b13953c5fb08fdb2b39635c1c8992 i added flag --no-timeout to ejabberdctl that will disable that timeout.

**Comment by mathiasertl** [comment]
Using the patch I can finally make complete dumps again - thank you.

Not really related to the original ticket, dumping is horribly slow. I posted some benchmark above already. The dump frequently stays at the same size for up to two and a half minutes only to slowly resume. The whole process takes around 22 minutes for what is in the end a ~1.6GB dump. Isn't there a way to speed this up?

**Comment by cromain** [comment]
closing this issue now as --no-timeout option will be available in 16.01
